From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Max Filippov <jcmvbkbc@gmail.com>
Date: Tue, 15 Aug 2023 11:24:43 -0700
Subject: [PATCH 22/73] drivers/mtd: add esp32 MTD support

Add FLASH mapping driver that uses direct mapping and IPC interface to
send erase/write commands to firmware.

Add esp32 partition reader so that instead of keeping synchronized copy
of the partition table in the DTS it could be read directly from FLASH.

Signed-off-by: Max Filippov <jcmvbkbc@gmail.com>
(cherry picked from commit 6e5dba46d921c050b6ce4732ca16508cca9e6790)
---
 drivers/mtd/chips/Kconfig     |   9 +-
 drivers/mtd/chips/Makefile    |   1 +
 drivers/mtd/chips/map_esp32.c | 402 ++++++++++++++++++++++++++++++++++
 drivers/mtd/maps/Kconfig      |   2 +-
 4 files changed, 412 insertions(+), 2 deletions(-)
 create mode 100644 drivers/mtd/chips/map_esp32.c

diff --git a/drivers/mtd/chips/Kconfig b/drivers/mtd/chips/Kconfig
index 19726ebd973d0..914f4dd16f477 100644
--- a/drivers/mtd/chips/Kconfig
+++ b/drivers/mtd/chips/Kconfig
@@ -233,11 +233,18 @@ config MTD_ABSENT
 
 config MTD_XIP
 	bool "XIP aware MTD support"
-	depends on !SMP && (MTD_CFI_INTELEXT || MTD_CFI_AMDSTD) && ARCH_MTD_XIP
+	depends on !SMP && (MTD_CFI_INTELEXT || MTD_CFI_AMDSTD || MTD_ESP32) && ARCH_MTD_XIP
 	default y if XIP_KERNEL
 	help
 	  This allows MTD support to work with flash memory which is also
 	  used for XIP purposes.  If you're not sure what this is all about
 	  then say N.
 
+config MTD_ESP32
+	tristate "Support for ESP32 FLASH chips in bus mapping"
+	select ESP32_IPC
+	help
+	  This option enables basic support for SPI FLASH chips accessed through
+	  a bus mapping driver and controlled through the ESP32 IPC mechanism.
+
 endmenu
diff --git a/drivers/mtd/chips/Makefile b/drivers/mtd/chips/Makefile
index 1f4e84f1cd88b..d54714d032c4e 100644
--- a/drivers/mtd/chips/Makefile
+++ b/drivers/mtd/chips/Makefile
@@ -14,3 +14,4 @@ obj-$(CONFIG_MTD_JEDECPROBE)	+= jedec_probe.o
 obj-$(CONFIG_MTD_RAM)		+= map_ram.o
 obj-$(CONFIG_MTD_ROM)		+= map_rom.o
 obj-$(CONFIG_MTD_ABSENT)	+= map_absent.o
+obj-$(CONFIG_MTD_ESP32)		+= map_esp32.o
diff --git a/drivers/mtd/chips/map_esp32.c b/drivers/mtd/chips/map_esp32.c
new file mode 100644
index 0000000000000..fabf6e0f4a5e5
--- /dev/null
+++ b/drivers/mtd/chips/map_esp32.c
@@ -0,0 +1,402 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#include <linux/completion.h>
+#include <linux/errno.h>
+#include <linux/esp32-ipc-api.h>
+#include <linux/init.h>
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/mtd/map.h>
+#include <linux/mtd/mtd.h>
+#include <linux/mtd/partitions.h>
+#include <linux/mtd/xip.h>
+#include <linux/of.h>
+#include <linux/of_platform.h>
+#include <linux/platform_device.h>
+#include <linux/slab.h>
+#include <linux/types.h>
+
+static int map_esp32_read(struct mtd_info *, loff_t, size_t, size_t *, u_char *);
+static int map_esp32_write(struct mtd_info *, loff_t, size_t, size_t *, const u_char *);
+static void map_esp32_nop(struct mtd_info *);
+static struct mtd_info *map_esp32_probe(struct map_info *map);
+static int map_esp32_erase(struct mtd_info *mtd, struct erase_info *info);
+static int map_esp32_point(struct mtd_info *mtd, loff_t from, size_t len,
+			 size_t *retlen, void **virt, resource_size_t *phys);
+static int map_esp32_unpoint(struct mtd_info *mtd, loff_t from, size_t len);
+
+enum {
+	ESP32_IPC_FLASH_STATE_IDLE,
+	ESP32_IPC_FLASH_STATE_READY,
+	ESP32_IPC_FLASH_STATE_DONE,
+};
+
+enum {
+	ESP_IPC_FLASH_CMD_ERASE,
+	ESP_IPC_FLASH_CMD_READ,
+	ESP_IPC_FLASH_CMD_WRITE,
+};
+
+struct esp32_ipc_flash_cmd {
+	void *data;
+	u32 addr;
+	u32 size;
+	u32 local_state;
+
+	u32 remote_state;
+	u32 result;
+	u32 code;
+};
+
+struct esp32_ipc_flash {
+	void *ipc;
+	u32 ipc_addr;
+	struct map_info *map;
+	struct esp32_ipc_flash_cmd *cmd;
+	struct completion *completion;
+};
+
+static int esp32_ipc_flash_erase(struct esp32_ipc_flash *hw, u32 off, u32 size);
+static int esp32_ipc_flash_read(struct esp32_ipc_flash *hw, u32 off, u32 size,
+				void *data);
+static int esp32_ipc_flash_write(struct esp32_ipc_flash *hw, u32 off, u32 size,
+				 const void *data);
+
+
+#define ESP32_PARTITION_TABLE_DEFAULT_ADDRESS	0x8000
+#define ESP32_PARTITION_RECORD_MAGIC0		0xaa
+#define ESP32_PARTITION_RECORD_MAGIC1		0x50
+
+struct esp32_partition_record {
+	u8 magic[2];
+	u8 type;
+	u8 subtype;
+	u32 offset;
+	u32 size;
+	char name[20];
+};
+
+static struct mtd_chip_driver map_esp32_chipdrv = {
+	.probe	= map_esp32_probe,
+	.name	= "map_rom",
+	.module	= THIS_MODULE
+};
+
+static unsigned int default_erasesize(struct map_info *map)
+{
+	const __be32 *erase_size = NULL;
+
+	erase_size = of_get_property(map->device_node, "erase-size", NULL);
+
+	return !erase_size ? map->size : be32_to_cpu(*erase_size);
+}
+
+static struct mtd_info *map_esp32_probe(struct map_info *map)
+{
+	struct mtd_info *mtd;
+	struct device_node *controller_node;
+	struct platform_device *controller_pdev;
+	struct esp32_ipc_flash *hw;
+
+	controller_node = of_parse_phandle(map->device_node, "controller", 0);
+	controller_pdev = of_find_device_by_node(controller_node);
+	if (!controller_pdev)
+		return NULL;
+	hw = platform_get_drvdata(controller_pdev);
+	hw->map = map;
+
+	mtd = kzalloc(sizeof(*mtd), GFP_KERNEL);
+	if (!mtd)
+		return NULL;
+
+	map->fldrv = &map_esp32_chipdrv;
+	mtd->priv = hw;
+	mtd->name = map->name;
+	mtd->type = MTD_ROM;
+	mtd->size = map->size;
+	mtd->_point = map_esp32_point;
+	mtd->_unpoint = map_esp32_unpoint;
+	mtd->_read = map_esp32_read;
+	mtd->_write = map_esp32_write;
+	mtd->_sync = map_esp32_nop;
+	mtd->_erase = map_esp32_erase;
+	mtd->flags = MTD_WRITEABLE;
+	mtd->erasesize = default_erasesize(map);
+	mtd->writesize = 1;
+	mtd->writebufsize = 1;
+
+	__module_get(THIS_MODULE);
+	return mtd;
+}
+
+
+static int map_esp32_point(struct mtd_info *mtd, loff_t from, size_t len,
+			   size_t *retlen, void **virt, resource_size_t *phys)
+{
+	struct esp32_ipc_flash *hw = mtd->priv;
+	struct map_info *map = hw->map;
+
+	if (!map->virt)
+		return -EINVAL;
+	*virt = map->virt + from;
+	if (phys)
+		*phys = map->phys + from;
+	*retlen = len;
+	return 0;
+}
+
+static int map_esp32_unpoint(struct mtd_info *mtd, loff_t from, size_t len)
+{
+	return 0;
+}
+
+static int map_esp32_read(struct mtd_info *mtd, loff_t from, size_t len,
+			  size_t *retlen, u_char *buf)
+{
+	struct esp32_ipc_flash *hw = mtd->priv;
+	struct map_info *map = hw->map;
+
+	map_copy_from(map, buf, from, len);
+	*retlen = len;
+	return 0;
+}
+
+static void map_esp32_nop(struct mtd_info *mtd)
+{
+	/* Nothing to see here */
+}
+
+static int map_esp32_write(struct mtd_info *mtd, loff_t to, size_t len,
+			   size_t *retlen, const u_char *buf)
+{
+	struct esp32_ipc_flash *hw = mtd->priv;
+	int ret;
+
+	ret = esp32_ipc_flash_write(hw, to, len, buf);
+	*retlen = len;
+	return ret;
+}
+
+static int map_esp32_erase(struct mtd_info *mtd, struct erase_info *info)
+{
+	struct esp32_ipc_flash *hw = mtd->priv;
+
+	return esp32_ipc_flash_erase(hw, info->addr, info->len);
+}
+
+static void __xipram esp32_ipc_flash_io_iram(struct esp32_ipc_flash_cmd *cmd)
+{
+	unsigned long flags;
+
+	local_irq_save(flags);
+	WRITE_ONCE(cmd->local_state, ESP32_IPC_FLASH_STATE_READY);
+	wmb();
+	while (READ_ONCE(cmd->remote_state) != ESP32_IPC_FLASH_STATE_DONE)
+		rmb();
+	local_irq_restore(flags);
+}
+
+static int esp32_ipc_flash_io(struct esp32_ipc_flash *hw,
+			      struct esp32_ipc_flash_cmd *cmd)
+{
+	int ret;
+
+	WRITE_ONCE(cmd->remote_state, ESP32_IPC_FLASH_STATE_IDLE);
+	WRITE_ONCE(cmd->local_state, ESP32_IPC_FLASH_STATE_IDLE);
+	wmb();
+
+	ret = esp32_ipc_tx(hw->ipc, hw->ipc_addr, cmd, NULL);
+	if (ret)
+		return ret;
+
+	while (READ_ONCE(cmd->remote_state) == ESP32_IPC_FLASH_STATE_IDLE)
+		rmb();
+	if (READ_ONCE(cmd->remote_state) != ESP32_IPC_FLASH_STATE_READY)
+		return -EINVAL;
+	esp32_ipc_flash_io_iram(cmd);
+	return cmd->result ? -EINVAL : 0;
+}
+
+static int esp32_ipc_flash_erase(struct esp32_ipc_flash *hw, u32 off, u32 size)
+{
+	hw->cmd->code = ESP_IPC_FLASH_CMD_ERASE;
+	hw->cmd->addr = off;
+	hw->cmd->size = size;
+	hw->cmd->data = NULL;
+	return esp32_ipc_flash_io(hw, hw->cmd);
+}
+
+static int esp32_ipc_flash_read(struct esp32_ipc_flash *hw, u32 off, u32 size,
+				void *data)
+{
+	hw->cmd->code = ESP_IPC_FLASH_CMD_READ;
+	hw->cmd->addr = off;
+	hw->cmd->size = size;
+	hw->cmd->data = data;
+	return esp32_ipc_flash_io(hw, hw->cmd);
+}
+
+static int esp32_ipc_flash_write(struct esp32_ipc_flash *hw, u32 off, u32 size,
+				 const void *data)
+{
+	hw->cmd->code = ESP_IPC_FLASH_CMD_WRITE;
+	hw->cmd->addr = off;
+	hw->cmd->size = size;
+	hw->cmd->data = (void *)data;
+	return esp32_ipc_flash_io(hw, hw->cmd);
+}
+
+static int esp32_ipc_flash_rx(void *p, void *data)
+{
+	struct esp32_ipc_flash *hw = p;
+
+	hw->cmd = data;
+	complete(hw->completion);
+	return 1;
+}
+
+static int esp32_ipc_flash_probe(struct platform_device *pdev)
+{
+	struct esp32_ipc_flash *hw =
+		devm_kzalloc(&pdev->dev, sizeof(*hw), GFP_KERNEL);
+	DECLARE_COMPLETION_ONSTACK(completion);
+	int ret;
+
+	if (!hw)
+		return -ENOMEM;
+	platform_set_drvdata(pdev, hw);
+
+	init_completion(&completion);
+	hw->completion = &completion;
+	hw->ipc = platform_get_drvdata(to_platform_device(pdev->dev.parent));
+	ret = of_property_read_u32(pdev->dev.of_node, "reg", &hw->ipc_addr);
+	if (ret < 0)
+		return ret;
+
+	ret = esp32_ipc_register_rx(hw->ipc, hw->ipc_addr, hw,
+				    esp32_ipc_flash_rx, NULL);
+	if (ret < 0)
+		return ret;
+
+	while (esp32_ipc_tx(hw->ipc, hw->ipc_addr, NULL, NULL) != 0) {
+	}
+	wait_for_completion(&completion);
+	if (!hw->cmd)
+		return -ENOMEM;
+	return 0;
+}
+
+static void esp32_ipc_flash_remove(struct platform_device *pdev)
+{
+}
+
+static const struct of_device_id esp32_ipc_flash_match[] = {
+	{
+		.compatible = "esp,esp32-ipc-flash",
+	}, {},
+};
+MODULE_DEVICE_TABLE(of, esp32_ipc_flash_match);
+
+static struct platform_driver esp32_ipc_flash_driver = {
+	.probe   = esp32_ipc_flash_probe,
+	.remove  = esp32_ipc_flash_remove,
+	.driver  = {
+		.name = "esp32-ipc-flash",
+		.of_match_table = of_match_ptr(esp32_ipc_flash_match),
+	},
+};
+
+static const struct of_device_id mtd_parser_esp32_of_match_table[] = {
+	{
+		.compatible = "esp,esp32-partition-table",
+	}, {},
+};
+MODULE_DEVICE_TABLE(of, mtd_parser_esp32_of_match_table);
+
+static int esp32_flash_parse_partitions(struct mtd_info *mtd,
+					const struct mtd_partition **pparts,
+					struct mtd_part_parser_data *data)
+{
+	struct esp32_ipc_flash *hw = mtd->priv;
+	struct esp32_partition_record pr[16];
+	u32 addr = ESP32_PARTITION_TABLE_DEFAULT_ADDRESS;
+	struct device_node *of_node;
+	int ret;
+
+	of_node = mtd_get_of_node(mtd);
+	if (of_node) {
+		struct device_node *part = of_get_child_by_name(of_node, "partitions");
+
+		if (part) {
+			of_property_read_u32(part, "esp32-partition-table-addr", &addr);
+			of_node_put(part);
+		}
+	}
+
+	ret = esp32_ipc_flash_read(hw, addr, sizeof(pr), pr);
+	if (!ret) {
+		int i;
+		struct mtd_partition *mtd_part = kzalloc(sizeof(*mtd_part) * ARRAY_SIZE(pr),
+							 GFP_KERNEL);
+
+		if (!mtd_part)
+			return -ENOMEM;
+
+		*pparts = mtd_part;
+
+		for (i = 0; i < ARRAY_SIZE(pr); ++i) {
+			struct esp32_partition_record *p = pr + i;
+			struct mtd_partition *mp = mtd_part + i;
+
+			if (p->magic[0] != ESP32_PARTITION_RECORD_MAGIC0 ||
+			    p->magic[1] != ESP32_PARTITION_RECORD_MAGIC1)
+				break;
+			mp->size = p->size;
+			mp->offset = p->offset;
+			mp->name = kstrdup(p->name, GFP_KERNEL);
+		}
+		ret = i;
+	}
+	return ret;
+}
+
+static void esp32_flash_cleanup(const struct mtd_partition *pparts, int nr_parts)
+{
+	int i;
+
+	for (i = 0; i < nr_parts; ++i)
+		kfree(pparts[i].name);
+	kfree(pparts);
+}
+
+static struct mtd_part_parser esp32_flash_partition_parser = {
+	.parse_fn = esp32_flash_parse_partitions,
+	.cleanup = esp32_flash_cleanup,
+	.name = "esp32",
+	.of_match_table = mtd_parser_esp32_of_match_table,
+};
+
+static int __init map_esp32_init(void)
+{
+	int ret = platform_driver_register(&esp32_ipc_flash_driver);
+
+	if (!ret) {
+		register_mtd_chip_driver(&map_esp32_chipdrv);
+		ret = register_mtd_parser(&esp32_flash_partition_parser);
+	}
+	return ret;
+}
+
+static void __exit map_esp32_exit(void)
+{
+	deregister_mtd_parser(&esp32_flash_partition_parser);
+	unregister_mtd_chip_driver(&map_esp32_chipdrv);
+	platform_driver_unregister(&esp32_ipc_flash_driver);
+}
+
+module_init(map_esp32_init);
+module_exit(map_esp32_exit);
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Max Filippov <jcmvbkbc@gmail.com>");
+MODULE_DESCRIPTION("MTD chip driver for ESP32 chips");
diff --git a/drivers/mtd/maps/Kconfig b/drivers/mtd/maps/Kconfig
index f447902d707e1..39a784bcb37b3 100644
--- a/drivers/mtd/maps/Kconfig
+++ b/drivers/mtd/maps/Kconfig
@@ -11,7 +11,7 @@ config MTD_COMPLEX_MAPPINGS
 
 config MTD_PHYSMAP
 	tristate "Flash device in physical memory map"
-	depends on MTD_CFI || MTD_JEDECPROBE || MTD_ROM || MTD_RAM || MTD_LPDDR
+	depends on MTD_CFI || MTD_JEDECPROBE || MTD_ROM || MTD_RAM || MTD_LPDDR || MTD_ESP32
 	help
 	  This provides a 'mapping' driver which allows the NOR Flash and
 	  ROM driver code to communicate with chips which are mapped
